internal/core/component_factory.rs

```diff
@@ -0,0 +1,287 @@
+        pub fn build(&self) -> Option<Component> {
+            let result = unsafe { (self.build)(self.data) };
+            if result.is_null() {
+                None
+            } else {
+                let r = unsafe { (*result).clone_internal() };
+                Some(r)
+            }
+        }
+    }
+
+    impl Default for ComponentFactoryData {
+        fn default() -> Self {
+            let data = Box::new(RustData { function: Box::new(|| None) });
+            Self {
+                build: __slint_component_factory_builder_call,
+                data: Box::into_raw(data) as *mut c_void,
+                drop: __slint_component_factory_builder_drop,
+            }
+        }
+    }
+
+    #[allow(unsafe_code)]
+    impl Drop for ComponentFactoryData {
+        fn drop(&mut self) {
+            unsafe { (self.drop)(self.data) };
+        }
+    }
+
+    pub(crate) struct RustData {
+        pub(crate) function: Box<dyn FnMut() -> Option<Component>>,
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn __slint_component_factory_builder_drop(data: *mut c_void) {
+        drop(Box::from_raw(data as *mut RustData));
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn __slint_component_factory_builder_call(
+        data: *mut c_void,
+    ) -> *mut Component {
+        let data = (data as *mut RustData).as_mut().unwrap();
+        if let Some(result) = (data.function)() {
+            Box::into_raw(Box::new(result))
+        } else {
+            core::ptr::null_mut()
+        }
+    }
+
+    pub(crate) struct ComponentFactoryInner(pub Rc<ComponentFactoryData>);
+}
+
+/// A Box containing a `ComponentFactory` (or nothing)
+#[repr(C)]
+pub struct ComponentFactory(core::ptr::NonNull<internal::ComponentFactoryInner>);
+
+impl Clone for ComponentFactory {
+    fn clone(&self) -> Self {
+        Self(
+            core::ptr::NonNull::new(Box::into_raw(Box::new(internal::ComponentFactoryInner(
+                self.factory().clone(),
+            ))))
+            .unwrap(),
+        )
+    }
+}
+
+impl Debug for ComponentFactory {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_tuple("ComponentFactory").finish()
+    }
+}
+
+impl Default for ComponentFactory {
+    fn default() -> Self {
+        Self(
+            core::ptr::NonNull::new(Box::into_raw(Box::new(internal::ComponentFactoryInner(
+                Rc::new(internal::ComponentFactoryData::default()),
+            ))))
+            .unwrap(),
+        )
+    }
+}
+
+#[allow(unsafe_code)]
+impl Drop for ComponentFactory {
+    fn drop(&mut self) {
+        drop(unsafe { Box::from_raw(self.0.as_ptr()) });
+    }
+}
+
+impl PartialEq for ComponentFactory {
+    fn eq(&self, other: &Self) -> bool {
+        Rc::ptr_eq(self.factory(), other.factory())
+    }
+}
+
+impl ComponentFactory {
+    /// Create a new `ComponentFactory`
+    #[allow(unsafe_code)]
+    pub fn new_raw(
```

Comment:
This must be marked as unsafe!  (anything that takes and use pointer usually must be unsafe)

Also, why is it public?